File: docs/snippets/all/archetypes/image_send_columns.rs
use ndarray::{s, Array, ShapeBuilder};
use rerun::Archetype as _;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec = rerun::RecordingStreamBuilder::new("rerun_example_image_send_columns").spawn()?;

    // Timeline on which the images are distributed.
    let times = (0..20).collect::<Vec<i64>>();

    // Create a batch of images with a moving rectangle.
    let width = 300;
    let height = 200;
    let mut images = Array::<u8, _>::zeros((times.len(), height, width, 3).f())
        .as_standard_layout() // Make sure the data is laid out as we expect it.
        .into_owned();
    images.slice_mut(s![.., .., .., 2]).fill(255);
    for &t in &times {
        let t = t as usize;
        images
            .slice_mut(s![t, 50..150, (t * 10)..(t * 10 + 100), 1])
            .fill(255);
    }

    // Log the ImageFormat and indicator once, as static.
    let format = rerun::components::ImageFormat::rgb8([width as _, height as _]);
    rec.log_static("images", &[&format as _, &rerun::Image::indicator() as _])?;

    // Split up the image data into several components referencing the underlying data.
    let image_size_in_bytes = width * height * 3;
    let blob = rerun::datatypes::Blob::from(images.into_raw_vec_and_offset().0);
    let image_column = times
        .iter()
        .map(|&t| {
            let byte_offset = image_size_in_bytes * (t as usize);
            rerun::components::ImageBuffer::from(
                blob.clone() // Clone is only a reference count increase, not a full copy.
                    .sliced(byte_offset..(byte_offset + image_size_in_bytes)),
            )
        })
        .collect::<Vec<_>>();

    // Send all images at once.
    let timeline_values = rerun::TimeColumn::new_sequence("step", times);
    rec.send_columns("images", [timeline_values], [&image_column as _])?;

    Ok(())
}

File: docs/snippets/all/archetypes/image_send_columns.cpp
// Send multiple images at once using `send_columns`.

#include <numeric>
#include <rerun.hpp>

int main() {
    auto rec = rerun::RecordingStream("rerun_example_image_send_columns");
    rec.spawn().exit_on_failure();

    // Timeline on which the images are distributed.
    std::vector<int64_t> times(20);
    std::iota(times.begin(), times.end(), 0);

    // Create a batch of images with a moving rectangle.
    const size_t width = 300, height = 200;
    std::vector<uint8_t> images(times.size() * height * width * 3, 0);
    for (size_t t = 0; t < times.size(); ++t) {
        for (size_t y = 0; y < height; ++y) {
            for (size_t x = 0; x < width; ++x) {
                size_t idx = (t * height * width + y * width + x) * 3;
                images[idx + 2] = 255; // Blue background
                if (y >= 50 && y < 150 && x >= t * 10 && x < t * 10 + 100) {
                    images[idx + 1] = 255; // Turquoise rectangle
                }
            }
        }
    }

    // Log the ImageFormat and indicator once, as static.
    auto format = rerun::components::ImageFormat(
        {width, height},
        rerun::ColorModel::RGB,
        rerun::ChannelDatatype::U8
    );
    rec.log_static("images", rerun::borrow(&format), rerun::Image::IndicatorComponent());

    // Split up the image data into several components referencing the underlying data.
    const size_t image_size_in_bytes = width * height * 3;
    std::vector<rerun::components::ImageBuffer> image_data(times.size());
    for (size_t i = 0; i < times.size(); ++i) {
        image_data[i] = rerun::borrow(images.data() + i * image_size_in_bytes, image_size_in_bytes);
    }

    // Send all images at once.
    rec.send_columns(
        "images",
        rerun::TimeColumn::from_sequence_points("step", std::move(times)),
        rerun::Image().with_many_buffer(std::move(image_data)).columns()
    );
}
